lib/Extension/LanguageServerPhpCsFixer/Provider/PhpCsFixerDiagnosticsProvider.php

```diff
@@ -0,0 +1,274 @@
+
+            $diagnostics = [];
+
+            $diffParser = new Parser();
+
+            foreach ($rules as $rule) {
+                $fileDiffText = yield $this->phpCsFixer->fix($textDocument->text, ['--dry-run', '--diff', '--rules', $rule]);
+                $fileDiff = $diffParser->parse($fileDiffText);
+
+                if (count($fileDiff) !== 1) {
+                    $this->logger->warning(
+                        sprintf("Expected php-cs-fixer to provide 1 diff, got %s. Skipping diagnostics for file '%s'", count($fileDiff), $textDocument->uri)
+                    );
+
+                    continue;
+                }
+
+                $fileDiff = $fileDiff[0];
+
+                foreach ($fileDiff->getChunks() as $chunk) {
+                    // diff is 1-indexed + in a line loop we update this number beforehand
+                    $lineNo = $chunk->getStart() - 2;
+
+                    /** @var Line[] */
+                    $changedLines = [];
+                    /** @var Line[]|null */
+                    $replacedLines = null;
+                    /** @var int|null */
+                    $startLineNo = null;
+
+                    foreach ($chunk->getLines() as $index => $line) {
+                        // increment orig file line number (added lines are not part of orig file)
+                        if (in_array($line->getType(), [Line::UNCHANGED, Line::REMOVED])) {
+                            $lineNo++;
+                        }
+
+                        $lastChangedLine = end($changedLines);
+
+                        // consume same as previous line
+                        if ($lastChangedLine && $line->getType() === $lastChangedLine->getType()) {
+                            $changedLines[] = $line;
+                            continue;
+                        }
+
+                        // consume lines if previous were removed and now we getting a replacement ones
+                        if ($lastChangedLine && $lastChangedLine->getType() === Line::REMOVED && $line->getType() === Line::ADDED) {
+                            $replacedLines = $changedLines;
+                            $changedLines = [$line];
+
+                            continue;
+                        }
+
+                        if ($lastChangedLine) {
+                            if (empty($changedLines) || !$startLineNo) {
+                                throw new LogicException("Missing data that's expected to be set");
+                            }
+
+                            $startPos = new Position($startLineNo, 0);
+                            $lineLength = strlen($lastChangedLine->getContent());
+                            $endPos = $lineLength
+                                ? new Position($lineNo - 1, $lineLength)
+                                : new Position($lineNo, 0);
+
+                            if ($replacedLines) {
+                                $firstLineA = $replacedLines[0]->getContent();
```

Comment:
this is a lot of indentation, can each block be separated into separate methods so that:

```
foreach ($foos as $bar) {
     $bars = $this->fooBars($bar, $boo, $baz); // can pass lines arrays by references
}
```

and make the variable scopes more explicit and perhaps make the code easier to understand/

Reply:
Extracted few methods which broke code into smaller readable or not repetitive tasks. Is that enough or is there a need for more?